e3/electrolyt/run.py

```diff
@@ -28,6 +34,90 @@ def run(self):
+        self.status = SUCCESS
+
+    def do_test(self):
+        """Run the test primitive."""
+        self.data.data.sandbox = self.sandbox
+        anod_driver = AnodDriver(self.data.data, None)
+        anod_driver.activate()
+        anod_driver.anod_instance.build_space.create(quiet=True)
+        if not hasattr(anod_driver.anod_instance, 'test'):
+            logger.error('primitive test not implemented')
+            return
+        anod_driver.anod_instance.test()
+        self.status = SUCCESS
+
+    def do_checkout(self):
+        """Get sources from Github to sbx/vcs."""
+        repo_name = self.data.data[0]
+        repo_url = self.data.data[1]['url']
+        repo_revision = self.data.data[1]['revision']
+        repo_vcs = self.data.data[1]['vcs']
+        if repo_vcs != 'git':
+            logger.error('%s vcs type not supported' % repo_vcs)
+            return
+        repo_dir = os.path.join(self.sandbox.vcs_dir, repo_name)
+        g = GitRepository(repo_dir)
+        if e3.log.default_output_stream is not None:
+            g.log_stream = e3.log.default_output_stream
+        g.init()
+        g.update(repo_url, repo_revision, force=True)
+        self.status = SUCCESS
+
+    def do_createsource(self):
+        """Copy src from vcs to cache using sourcebuilders."""
+        source_name = self.data.data[1]
+        tmp_cache_dir = os.path.join(self.sandbox.tmp_dir, 'cache')
+        src = self.sandbox.vcs_dir
+        for src_builder in self.data.data[0].source_pkg_build:
+            if src_builder.name == source_name:
+                repo_dict = {}
+                src_dir = os.path.join(src, src_builder.checkout[0])
+                dest_dir = os.path.join(tmp_cache_dir, src_builder.name)
+                repo_dict[src_builder.name] = {"working_dir": src_dir}
+                src_builder.prepare_src(repo_dict, dest_dir)
+                self.status = SUCCESS
+                logger.debug('%s created in cache/tmp' % source_name)
+                return
+
+    def do_getsource(self):
+        """action_item from an intermediate node."""
+        self.status = SUCCESS
+
+    def do_installsource(self):
+        """Install the source from tmp/cache to build_space/src."""
+        spec = self.data.data[0]
+        spec.sandbox = self.sandbox
+        anod_instance = AnodDriver(spec, None)
+        anod_instance.activate()
+        source = self.data.data[1]
+        src_dir = os.path.join(self.sandbox.tmp_dir,
+                               'cache',
+                               source.name)
+        dest_dir = os.path.join(spec.build_space.src_dir)
+        sync_tree(src_dir, dest_dir, ignore=source.ignore)
+        self.status = SUCCESS
+
+    def do_root(self):
+        """Epress the final result of the exec."""
+        # If nothing fails in the process the status will remain automatically
+        # UKNOWN, as the root node UNKNOWN status is a SUCCESS
+        if self.status == STATUS_UNKNOWN:
```

Comment:
the status of the job is unknown by default if none of its predecessors has failed